src/index.ts

```diff
@@ -211,6 +219,32 @@ function newFactory(config: Config, interfaceDefaults: Record<string, string>, t
   return [factory, ...(isEnumDetailObject(type) ? [] : [maybeFunctions])];
 }
 
+/** Creates a `new${type}` function for the given `type`. */
+function newInterfaceFactory(config: Config, interfaceName: string, impls: string[]): Code[] {
+  const defaultImpl = impls[0] || fail(`Interface ${interfaceName} is unused`);
+  return [
+    code`
+      type ${interfaceName}Options = ${impls.map(name => `${name}Options`).join(" | ")};
+    `,
+
+    code`
+      type ${interfaceName}Type = ${impls.join(" | ")};
+    `,
+
+    code`
+      function maybeNew${interfaceName}(value: ${interfaceName}Options | undefined, cache: Record<string, any>): ${interfaceName}Type {
```

Comment:
Cool, fwiw we might need a `maybeNew{...}OrNull` for nullable interfaces.

Reply:
Added `OrNull` variant